### PR TITLE
Nested modules

### LINKED ISSUE #543: Allow nested suites (modules)?
A personal sore point for me when using QUnit is the lack of nested suites (`module`s) that are offered by other popular frameworks like Mocha and Jasmine.

I've given it some cursory thought and, if we agree it adds value, I don't foresee any major roadblocks to implementing that capability.  It would give us some options for richer reporting outputs as well, e.g. Mocha's sleek nested HTML report:

screen shot 2014-03-12 at 1 00 06 pm

### PR DESCRIPTION
**Pull request by leobalter**
Replaces/Closes #800 
Closes #543

This PR closes #800 with extra additions to the PR to assure a good landing of nested modules.

We might want to observe the issues listed at #858 and add extra commits

cc @gibson042

### CODE REVIEW

File: test/main/modules.js [hunk 1]
```diff
@@ -147,3 +147,57 @@ QUnit.module( "Deprecated setup/teardown", {
...
+	QUnit.test( "test after all nesting modules processed and new module declared",
+		function( assert ) {
+			var module = assert.test.module;
+			assert.equal( module.name, "nested modules > second outer" );
+		});
+});
+
+QUnit.test( "modules with nested functions does not spread beyond", function( assert ) {
+	assert.equal( assert.test.module.name, "pre-nested modules" );
+});
```
Comment (leobalter) [inline test/main/modules.js]: @jzaefferer @gibson042 I need opinions on this test.

Comment (gibson042) [inline test/main/modules.js]: Seems ok to me. The interaction itself sucks, but the test captures it well.

**Comment by leobalter** [comment]
@gibson042 I made two commits to include arguments on the contained modules callback. I appreciate if you review and criticize.

cc @jzaefferer

File: test/main/modules.js [hunk 2]
```diff
@@ -147,3 +147,102 @@ QUnit.module( "Deprecated setup/teardown", {
...
+		assert.strictEqual( typeof hooks.afterEach, "function" );
+	} );
+
+	QUnit.module( "contained suite argument hooks", function( test, hooks ) {
+		var beforeEach = hooks.beforeEach;
+		var afterEach = hooks.afterEach;
+
+		beforeEach( function( assert ) {
+			assert.ok( true, "beforeEach called" );
+		} );
```
Comment (leobalter) [inline test/main/modules.js]: This is even better on ES6:

```js
QUnit.module( "foo", ( test, { beforeEach, afterEach } ) => {
  beforeEach: () => {
    ...
  };

  ...
});
```

Also, a funny fact:

```js
const describe = QUnit.module;

describe( "wat", it => {
  it( "works", ( { ok } ) => {
    ok( true, "omg" );
  });
});
```

File: src/core.js_ [hunk 3]
```diff
@@ -0,0 +1,312 @@
+QUnit.urlParams = urlParams;
```
Comment (gibson042) [inline src/core.js_]: I'm assuming these `.js_` files are accidental additions.

Comment (leobalter) [inline src/core.js_]: caught in a git add .

Comment (leobalter) [inline src/core.js_]: 57527a3

**Comment by gibson042** [comment]
> I made two commits to include arguments on the contained modules callback. I appreciate if you review and criticize.

I'm not sure about these particular parameters. As I've noted [elsewhere](https://github.com/jquery/qunit/pull/670#discussion_r26321413), I'd like `module` and `test` callback signatures to look as similar as possible. But further than that, I'd like `module` callbacks to be straightforward. It seems to me that data passed in to them—including as `this`—should be entirely sufficient to define submodules/tests/etc., but hopefully not exceed that.

In my opinion, it's better to pass nothing at all now than it is to guess wrong about the signature. Since this iteration still requires access to the global `QUnit` for some operations (e.g., submodules), I'd actually prefer an argument-free `executeNow()` for this PR, saving context/arguments for later [as discussed](https://github.com/jquery/qunit/pull/800#issuecomment-139336531). For example, a possible future that would be precluded by passing `test` now:
```js
QUnit.module("outer", function( outerModule ) {
	// environment (prototype for test/submodule context)
	this.outer = true;
	this.outerBefore = 0;

	// test before/after
	outerModule.test.beforeEach(function() {
		this.outerBefore++;
	});

	// tests
	outerModule.test("foo", function( assert ) {
		assert.equal(this.outerBefore, 1, "beforeEach updated environment");
	});
	outerModule.test("bar", function( assert ) {
		assert.equal(this.outerBefore, 1, "sibling tests don't share environments");
	});

	// submodules
	outerModule.module("inner", function( innerModule ) {
		// environment (inherits from parent)
		this.inner = true;

		function assertEnvironment( assert, before ) {
			assert.ok(this.outer, "outer module initialized environment");
			assert.ok("outerBefore" in this, "outer module beforeEach updated environment");
			assert.equal(this.outerBefore, 1, "submodules don't share environments");
			assert.ok(this.inner, "inner module initialized environment");
			if ( before ) {
				return;
			}
			assert.ok(this.innerBefore, "inner module beforeEach updated environment");
			assert.ok(this.tested, "test updated environment");
		}

		// test before/after
		innerModule.test.beforeEach(function( assert ) {
			this.innerBefore = true;
			assertEnvironment( assert, true );
		});
		innerModule.test.afterEach(function( assert ) {
			assertEnvironment( assert );
		});

		// tests
		innerModule.test("bar", function( assert ) {
			this.tested = true;
			assertEnvironment( assert );
		});
	});
});
```

**Comment by leobalter** [comment]
```js
QUnit.module("outer", function( outerModule ) {
    // environment (prototype for test/submodule context)
    this.outer = true;
    this.outerBefore = 0;

    // test before/after
    outerModule.test.beforeEach(function() {
        this.outerBefore++;
    });

    // tests
    outerModule.test("foo", function( assert ) {
        assert.equal(this.outerBefore, 1, "beforeEach updated environment");
    });
    outerModule.test("bar", function( assert ) {
        assert.equal(this.outerBefore, 1, "sibling tests don't share environments");
    });

    ...
```

These `this.outerBefore` tests seems very confusing to me. I can understand the point here but I'm uncomfortable with it so far.

---

> I'm not sure about these particular parameters.

And reviewing my code I'll indeed remove the first `test` argument.

And following your own example, I would stick with the first argument being an object with the hooks methods and it can later being extended. This way:

```js
QUnit.module("outer", function( outerModule ) {
    ...
    outerModule.beforeEach(function() {
        ...
    });

    outerModule.afterEach(function() {
        ...
    });
```

but also, later in further additions:

```js
QUnit.module("outer", function( t ) {
    t.beforeEach(function() {
        ...
    });
    t.test( "the test method", function( assert ) {
        ...
    });
});
```

and using ES6:

```js
QUnit.module("outer", ( { beforeEach } ) => {
    beforeEach(function() {
        ...
    });
    QUnit.test( "the test method", function( assert ) {
        ...
    });
});
```

The reason to start with `beforeEach` and `afterEach` methods is to avoid calling modules with three arguments, where the hooks object is already big enough:

```js
QUnit.module( "outer", {
        beforeEach: function( assert ) {
            ...
        },
        afterEach: function( assert ) {
            ...
        }
    },
    function( t ) {
        QUnit.test( "the test method", function( assert ) {
            ...
        });
    }
);
```

**Comment by leobalter** [comment]
@gibson042: both 29f5b84 and 53778e3 has the shared `this` based on your example.

The `this` object is only propagated from the outer to the inner scopes without duplicating the test hooks.

I also removed the test argument as you can see on the fixup commits.

We may then extend the module callback argument to include `test` later. I want to at least bring the shared `this` and the hooks when landing the nested modules.

File: src/core.js [hunk 4]
```diff
@@ -27,8 +31,51 @@ extend( QUnit, {
...
+				[ parentModule.name, name ].join( " > " ) : name;
+			var module = {
+				name: moduleName,
+				parentModule: parentModule,
+				tests: []
+			};
+
+			var env = {};
+			if ( parentModule ) {
+				extend( env, parentModule.testEnvironment );
```
Comment (gibson042) [inline src/core.js]: I think `Object.create`-style inheritance would be better, but am willing to shelve that for now.

File: test/main/modules.js [hunk 5]
```diff
@@ -147,3 +147,141 @@ QUnit.module( "Deprecated setup/teardown", {
...
+
+QUnit.module( "contained suite `this`", function( hooks ) {
+	this.outer = 1;
+
+	hooks.beforeEach( function() {
+		this.outer++;
+	} );
+
+	hooks.afterEach( function( assert ) {
+		assert.equal( this.outer, 2 );
```
Comment (gibson042) [inline test/main/modules.js]: The increment-based tests in https://github.com/leobalter/qunit/commit/29f5b842e61e2b735b0cc9046034c26409375271#diff-3ff526461245427b5e9714efeb6ca4ed failed, right? It is unfortunate, but IIRC tests within a module _do_ currently share environment.

**Comment by gibson042** [comment]
:+1: The single callback argument is a decent (if limited) module stand-in and has good room for growth, and you are absolutely right about this:
> The reason to start with `beforeEach` and `afterEach` methods is to avoid calling modules with three arguments, where the hooks object is already big enough

Let's hope that `module` and `test` follow suit, so that we can eventually get to a point where `module` callbacks need nothing other than what's passed directly to them, just like current `test` callbacks.

File: src/core.js (hunk 4 above)
Comment (leobalter) [inline src/core.js]: I would love to do this. We still support ES3 until QUnit 2.0.0

File: test/main/modules.js (hunk 5 above)
Comment (leobalter) [inline test/main/modules.js]: this commit will be merged with 53778e3.

It is passing here. I just wanted to set the steps on different commits during the review stage.

> It is unfortunate, but IIRC tests within a module do currently share environment.

I've done it this way getting from this assertion below from your example.

```js
assert.ok(this.outer, "outer module initialized environment");
```

Should it be different?

Comment (gibson042) [inline test/main/modules.js]: Turns out the environments _are_ already distinct (:raised_hands:), so there's just some missing code here to show that in-test environment modifications are visible by `afterEach` callbacks but not by subsequent tests (as demonstrated in https://github.com/leobalter/qunit/commit/29f5b842e61e2b735b0cc9046034c26409375271#diff-3ff526461245427b5e9714efeb6ca4ed ).

File: src/core.js (hunk 4 above)
Comment (gibson042) [inline src/core.js]: Arbitrary inheritance is possible in ES3, too:
```js
var createObject = (function( Base ) {
	return function( prototype ) {
		Base.prototype = prototype;
		return new Base();
	};
})( Function() );
…
env = createObject( parentModule.testEnvironment );
```

The above is similar to a stripped-down version of lodash's [`baseCreate`](https://github.com/lodash/lodash/blob/d407d8aa08a6f739441a82e3b7c9d8a5cf66ba16/lodash.js#L1987-L2005), but you could easily go all the way to [`_.create`](https://lodash.com/docs#create) if you want this part to look like `module.testEnvironment = createObject( parentModule.testEnvironment, testEnvironment )` (removing the need for a separate `extend`).

File: test/main/modules.js (hunk 5 above)
Comment (leobalter) [inline test/main/modules.js]: Good catch. I got it on ce70ce9

File: src/core.js (hunk 4 above)
Comment (leobalter) [inline src/core.js]: This is cool, but I would leave this improvement for later as this requires a whole new review. Does it works for you?

Comment (gibson042) [inline src/core.js]: Yep.

**Comment by gibson042** [comment]
I think this is good to go; [module environment inheritance](https://github.com/jquery/qunit/pull/859#discussion_r40965697) can be addressed with a subsequent issue & PR.

**Comment by leobalter** [comment]
Opened #869 to follow up with this.

**Pull request closed without merging**